### ISSUE TITLE
Add support for package.json "browser" field

### ISSUE THREAD

**Issue by cyparu** [jspm/jspm-cli#1062]
Consider to add support for [browser](https://gist.github.com/defunctzombie/4339901) field spec for **package.json**: 
>The ```browser``` field is provided by a module author as a hint to javascript bundlers or component tools when preparing modules for client side use.

This issue have been brought in discussion in context of *SystemJS/jspm ES6 loader support/plugin* marko-js/marko#134 issue.

**Comment by guybedford** [comment]
We do support the `browser` field for npm package installs, but jspm itself doesn't by default recognize this property. You can have a jspm-specific main via the package.json `jspm.main` property though.

**Comment by neothemachine** [comment]
Doesn't work for me. My [browser field](https://github.com/Reading-eScience-Centre/covjson-reader/blob/master/package.json) is:
```js
  "browser": {
    "./lib/http.js": "./lib/http-browser.js",
    "request-promise": false
  },
```
But JSPM doesn't apply the replacement of http.js with http-browser.js when depending on that library.

**Comment by guybedford** [comment]
@neothemachine the browser field is only respected for npm packages as it is an npm convention not a jspm one.

**Comment by neothemachine** [comment]
Well I depend on it in npm style, if that's what you mean. So in my jspm dependencies it says `npm:covjson-reader`. Or do you mean something else?

**Comment by neothemachine** [comment]
You can easily try it out in the jspm.io console:
```
System.import('npm:covjson-reader').then(function(CovJSON) {})
```
When you look at what's loaded in the network tab, you see that it's "http.js" (https://npm.jspm.io/covjson-reader@0.3.2/lib/http.js) and it is really the node version. You can also see that there's a funny require statement `a=require("@empty")` which probably corresponds to the `request-promise` dependency that is removed via the "browser" field. It seems the "browser" field is only partially applied.

**Comment by guybedford** [comment]
Interesting, that does sound like it could be a bug then.

**Comment by guybedford** [comment]
(But it is probably something a lot more specific here, the generic browser support definitely works)

**Comment by neothemachine** [comment]
Should I move this to a separate issue or do you want to reopen this one?

**Comment by neothemachine** [comment]
I just invoked jspm-npm 0.26.0 manually and the conversion works:
```js
var fs = require('graceful-fs');
var nodeConversion = require('../lib/node-conversion');
var NPMLocation = require('../lib/npm')
var npm = new NPMLocation({})

var dir = './test/covjson-reader'
var pjson = JSON.parse(fs.readFileSync(dir + '/package.json'))
pjson = npm.processPackageConfig(pjson, 'covjson-reader')

nodeConversion.convertPackage(pjson, dir).then(function (result) {
  console.log('done')
})
```
No idea if this is how JSPM uses it, but this changes the library folder files in-place on disk and correctly replaces `require('./http');` with `require('./http-browser');`. I just tried JSPM 0.16.20 to check, but it doesn't do the replacement in jspm_packages. I don't know how to debug things when running `jspm install`, I tried adding some `console.error` statements in `node_modules/jspm/node_modules/jspm-npm/lib/node-conversion.js` but they seem to get swallowed by jspm. Any other way to output something to the console?

**Comment by neothemachine** [comment]
I found the bug!! What I wrote in my last comment was actually only working because I copied the covjson-reader library files out of jspm_packages (as I thought they were not changed so it wouldn't matter if I used them or the npm tarball). But in fact jspm-npm had already changed them, and applying the conversion twice brought the desired effect.

They went from `require('./http.js')` to `require('./http')`, and on the second run to `require('./http-browser')`! So it is some kind of edge case bug when you require modules with file extension.
